painttiming.bs

```diff
@@ -62,9 +87,41 @@ Formally, we consider the user agent to have "rendered" a document when it has p
 
     NOTE: The rendering pipeline is very complex, and the timestamp should be the latest timestamp the user agent is able to note in this pipeline (best effort). Typically the time at which the frame is submitted to the OS for display is recommended for this API.
 
+A [=generated content pseudo-element=] is a <dfn>paintable pseudo-element</dfn> when all of the following apply:
+* The pseudo-element's [=used=] [=visibility=] is <code>visible</code>.
+* The pseudo-element's [=used=] [=opacity=] is greater than zero.
+
+An [=element=] is <dfn export>contentful</dfn> when one or more of the following apply:
+* The element has a [=text node=] child, with at least one character, excluding [=document white space characters=].
+* The element is a [=replaced element=] representing a [=loaded=] image.
+* The element's [=used=] style contains a [=loaded=] [=background-image=].
+
+    NOTE: a gradient is not considered a loaded background image, and thus would not make an element count as contentful.
+
+* The element is a [=canvas=] with its [=context mode=] set to any value other than <code>none</code>.
+* The element is an [=svg element with rendered descendants=].
+* The element is an [=originating element=] for a [=paintable pseudo-element=] that represents a [=loaded=] image or text with at least one character excluding [=document white space characters=].
+
+<dfn>Paintable Bounding Rect</dfn>: The paintable bounding rect is the [=element=]'s bounding rect, which would be the result of running the [=getBoundingClientRect=] algorithm on the [=element=], clipped by ancestors with <code>overflow: clip</code>.
+
+    NOTE: elements contained by boxes with <code>overflow: scroll</code> or <code>overflow: hidden</code> don't have their [=paintable bounding rect=] clipped, as in both cases the [=element=] can become visible by scrolling.
+
+An [=element=] is <dfn>paintable</dfn> when all of the following apply:
+* The element's [=used=] [=visibility=] is <code>visible</code>.
+* The element and all of its ancestors' [=used=] [=opacity=] is greater than zero.
+
+    NOTE: there could be cases where a <code>paintable</code> [=element=] would not be visible to the user, for example in the case of text that has the same color as its background.
+    Those elements would still considered as paintable for the purpose of computing [=first contentful paint=].    
+
+* The element's [=paintable bounding rect=] has positive right, bottom, width and height values. 
```

Comment:
I don't think this is right when the root layout box is RTL.
Also, [DOMRect](https://drafts.fxtf.org/geometry-1/#domrect) doesn't seem to define "right"?

Reply:
DOMRect does define right - but I'll fix the RTL thing.

Reply:
Hm... now that I'm thinking about this more. Not sure how this will interact with the margin tiles in WebKit.

@smfr: do we clip the document's contents at where the coordinates normally "ends"? Or do we end up painting onto the margin tiles?

Reply:
I personally still think that it's better if the spec was viewport-based - and the first-contentful-paint time would be equivalent to the first time an IntersectionObserver with the document as the root firing an event for any contentful/paintable element. Otherwise - maybe remove this requirement and allow elements that would never get painted affect FCP?

Reply:
For now revised it to "intersects with the scrolling area of the document", which is maybe more precise.